### PR TITLE
Allows clothing to alter variables on the wearer, Admin modifiable too!

### PR DESCRIPTION
**Pull request by RemieRichards**
- Allows clothing to modify the wearer's variables, any of them!
- The variables that clothing modify are COMPLETELY ADMIN EDITABLE!
- System requested by @Robustin (Plus I wanted to make a spooky ghost suit)
- Added a spooky ghost suit! Admin-only, Renames you to "Spooky Ghost", makes you partially transparent and lets you walk through anything!

spookysuit

### CODE REVIEW

**Comment by MMMiracles** [comment]
this is actually super cool and i'm surprised it wasn't already a thing

**Comment by PKPenguin321** [comment]
i mean techincally you can just adjust the mob's vars on equip() or equipped() or whatever but this is still cool

**Comment by RemieRichards** [comment]
@PKPenguin321 this DOES do it on equipped, it's an ease of use system (this way is also admin editable, which unique equipped()'s are not)

**Comment by Lumipharon** [comment]
Neat. One thing that comes to mind is tagging people with a particular faction, with like an armband or something. 
So you could have an away mission with a room full of super rape demons, but you can go somewhere else and find a mcguffin hat to make them not kill you when you go through.

Does this also work with accessories, or just the clothes themselves?

**Comment by RemieRichards** [comment]
@Lumipharon anything subtype'd from ```/clothing```, I have no idea about accessories' paths.

**Comment by Lumipharon** [comment]
RIP. Would it be possible to make clothes inherit the var changes of attached accessories, or something?

File: code/modules/clothing/suits/miscellaneous.dm [hunk 1]
```diff
@@ -490,3 +490,10 @@
 
 /obj/item/clothing/head/winterhood/miner
 	icon_state = "winterhood_miner"
+
+/obj/item/clothing/suit/spookyghost
+	name = "spooky ghost"
+	desc = "this is obviously just a bedsheet, but maybe try it on?"
+	icon_state = "bedsheet"
+	user_vars_to_edit = list("name" = "Spooky Ghost", "real_name" = "Spooky Ghost" , "incorporeal_move" = 1, "appearance_flags" = 288) //288 is KEEP_TOGETHER|TILE_BOUND
```
Comment (JJRcop) [inline code/modules/clothing/suits/miscellaneous.dm]: Why not just use 
```DM
"appearance_flags" = KEEP_TOGETHER|TILE_BOUND
```

Comment (RemieRichards) [inline code/modules/clothing/suits/miscellaneous.dm]: That would be because I had to work out the numbers on the fly to add it via VV because I was too lazy to recompile so I had the number 288 stuck in my head. I'll sort it in a commit tomorrow.

**Comment by coiax** [comment]
Is this going to lead to runtimes if the variable specified doesn't exist on the user? We're sidestepping protection of the compiler on this one.

File: code/modules/clothing/clothing.dm [hunk 2]
```diff
@@ -19,6 +19,37 @@
...
+	if(user_vars_remembered.len)
+		for(var/variable in user_vars_remembered)
+			if(variable in user.vars)
+				if(user.vars[variable] == user_vars_to_edit[variable]) //Is it still what we set it to? (if not we best not change it)
+					user.vars[variable] = user_vars_remembered[variable]
+		user_vars_remembered = list()
+
+
+/obj/item/clothing/equipped(mob/user, slot)
+	..()
```
Comment (phil235) [inline code/modules/clothing/clothing.dm]: did you check whether equipped() and dropped() need to return ..() ? I think I remember it being necessary.

Comment (RemieRichards) [inline code/modules/clothing/clothing.dm]: @phil235 I checked around the calls (obviously), and the calls for them aren't in locations where anything can read the return value, so if there ARE some, please point them out for me and I'll change it.

**Comment by RemieRichards** [comment]
@coiax No, please read the code.

File: code/__DEFINES/clothing.dm [hunk 3]
```diff
@@ -49,6 +37,49 @@
 
 #define slots_amt			20 // Keep this up to date!
 
+//I hate that this has to exist
```
Comment (coiax) [inline code/__DEFINES/clothing.dm]: Why does it exist then? I can't quite grasp the difference between slots and SLOTS

File: code/modules/clothing/clothing.dm [hunk 4]
```diff
@@ -19,6 +19,37 @@
...
+
+	if(slot_flags & slotdefine2slotbit(slot)) //Was equipped to a valid slot for this item?
```
Comment (coiax) [inline code/modules/clothing/clothing.dm]: Seems like we could replace this with another proc/macro

File: code/__DEFINES/clothing.dm (hunk 3 above)
Comment (RemieRichards) [inline code/__DEFINES/clothing.dm]: One is binary bit values, or "bitflags", the other is just raw numbers, this has to exist because our slot system is unbelievably shit, so clothing defines where it can be worn as a "bitfield" (a collection of "bitflags" stored in one number because yay binary), I have to convert the ```slot``` arg passed by ```equipped()``` to a bitflag appropriate for it's purpose (eg: slot_back to SLOT_BACK) to then see if it's in ONE of the clothing's slot flags. typically ```equipped()```s are written uniquely on each object so for example a ```/head``` item can just check if ```slot == slot_head``` and continue on, I implemented my system nice and generically at ```/clothing``` (for ease of use, this way you just define a list of vars to change and their values and it will all just -work-) but because of that genericness I can't simply check if slot is correct (clothing has NO concept of what slot it goes it, but it does know what slot FLAGS it can go in)

tl;dr it's the nicest workaround bar actually fixing the slot system, and good luck getting anyone to actually do that.

Comment (coiax) [inline code/__DEFINES/clothing.dm]: One is where it is, and one is where it fits? And the two have no relation  to each other?

File: code/modules/clothing/clothing.dm (hunk 2 above)
Comment (phil235) [inline code/modules/clothing/clothing.dm]: Oh yeah, sorry, I was misremembering the work I did on item actions. What was needed was just "..()".
